Frends.HTTP.DownloadFile/Frends.HTTP.DownloadFile/Extensions.cs

```diff
@@ -0,0 +1,129 @@
+                handler.UseDefaultCredentials = true;
+                break;
+            case Authentication.WindowsAuthentication:
+                var domainAndUserName = options.Username.Split('\\');
+                if (domainAndUserName.Length != 2)
+                    throw new ArgumentException($@"Username needs to be 'domain\username' now it was '{options.Username}'");
+
+                handler.Credentials =
+                    new NetworkCredential(domainAndUserName[1], options.Password, domainAndUserName[0]);
+                break;
+            case Authentication.ClientCertificate:
+                handler.ClientCertificates.AddRange(GetCertificates(options));
+                break;
+        }
+
+        handler.AllowAutoRedirect = options.FollowRedirects;
+        handler.UseCookies = options.AutomaticCookieHandling;
+
+        if (options.AllowInvalidCertificate)
+            handler.ServerCertificateCustomValidationCallback = (a, b, c, d) => true;
+    }
+
+    internal static void SetDefaultRequestHeadersBasedOnOptions(this HttpClient httpClient, Options options)
+    {
+        //Do not automatically set expect 100-continue response header
+        httpClient.DefaultRequestHeaders.ExpectContinue = false;
+        httpClient.DefaultRequestHeaders.TryAddWithoutValidation("content-type", "application/json");
+        httpClient.Timeout = TimeSpan.FromSeconds(Convert.ToDouble(options.ConnectionTimeoutSeconds));
+    }
+
+    private static X509Certificate[] GetCertificates(Options options)
+    {
+        X509Certificate2[] certificates;
+
+        switch (options.ClientCertificateSource)
+        {
+            case CertificateSource.CertificateStore:
+                var thumbprint = options.CertificateThumbprint;
+                certificates = GetCertificatesFromStore(thumbprint, options.LoadEntireChainForCertificate);
+                break;
+            case CertificateSource.File:
+                certificates = GetCertificatesFromFile(options.ClientCertificateFilePath, options.ClientCertificateKeyPhrase);
+                break;
+            case CertificateSource.String:
+                certificates = GetCertificatesFromString(options.ClientCertificateInBase64, options.ClientCertificateKeyPhrase);
+                break;
+            default:
+                throw new Exception("Unsupported Certificate source");
+        }
+
+        return certificates.Cast<X509Certificate>().ToArray();
+    }
+
+    private static X509Certificate2[] GetCertificatesFromString(string certificateContentsBase64, string keyPhrase)
+    {
+        var certificateBytes = Convert.FromBase64String(certificateContentsBase64);
+        return LoadCertificatesFromBytes(certificateBytes, keyPhrase);
+    }
+
+    // Have to skip these because there's no certificate to use for unit testing.
+    [ExcludeFromCodeCoverage]
+    private static X509Certificate2[] LoadCertificatesFromBytes(byte[] certificateBytes, string keyPhrase)
```

Comment:
You should be able to test this, since you can read the certificate as byte array. Your unit test helper method even creates the certificate byte array.